packages/api/src/router/stripe/index.ts

```diff
@@ -0,0 +1,141 @@
+import { z } from "zod";
+
+import { eq } from "@openstatus/db";
+import { user, usersToWorkspaces, workspace } from "@openstatus/db/src/schema";
+
+import { createTRPCRouter, protectedProcedure } from "../../trpc";
+import { stripe } from "./shared";
+import { webhookRouter } from "./webhook";
+
+const url =
+  process.env.NODE_ENV === "production"
+    ? "https://www.openstatus.dev"
+    : "http://localhost:3000";
+
+export const stripeRouter = createTRPCRouter({
+  webhooks: webhookRouter,
+
+  getUserCustomerPortal: protectedProcedure
+    .input(z.object({ workspaceId: z.string() }))
+    .mutation(async (opts) => {
+      const result = await opts.ctx.db
+        .select()
+        .from(workspace)
+        .where(eq(workspace.slug, opts.input.workspaceId))
+        .get();
+
+      if (!result) return;
+
+      const currentUser = opts.ctx.db
+        .select()
+        .from(user)
+        .where(eq(user.tenantId, opts.ctx.auth.userId))
+        .as("currentUser");
+      const userHasAccess = await opts.ctx.db
+        .select()
+        .from(usersToWorkspaces)
+        .where(eq(usersToWorkspaces.workspaceId, result.id))
+        .innerJoin(currentUser, eq(usersToWorkspaces.userId, currentUser.id))
+        .get();
+
+      if (!userHasAccess || !userHasAccess.users_to_workspaces) return;
+      let stripeId = result.stripeId;
+      if (!stripeId) {
+        const customerData: {
+          metadata: { workspaceId: string };
+          email?: string;
+        } = {
+          metadata: {
+            workspaceId: String(workspace.id),
+          },
+          email: opts.ctx.auth.user?.emailAddresses[0].emailAddress || "",
+        };
+
+        const stripeUser = await stripe.customers.create(customerData);
+
+        stripeId = stripeUser.id;
+        await opts.ctx.db
+          .update(workspace)
+          .set({ stripeId })
+          .where(eq(workspace.id, result.id))
+          .run();
+      }
+
+      const session = await stripe.billingPortal.sessions.create({
+        customer: stripeId || "",
+        return_url: `${url}/app/${result.slug}/settings`,
+      });
```

Comment:
As an idea for good UX, we could add a "thank you notification".

Reply:
Here it's just to update your info it's the customer portal it could be done in the checkout though